File: tests/test_v1_26_d_info_board_navigation.js
'use strict';

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const root = path.resolve(__dirname, '..');
const source = fs.readFileSync(path.join(root, 'public/js/info-board-navigation.js'), 'utf8');

let domReadyHandler = null;
const windowStub = {
    setTimeout: setTimeout,
    clearTimeout: clearTimeout,
    addEventListener: function () {}
};
const documentStub = {
    readyState: 'loading',
    addEventListener: function (name, handler) {
        if (name === 'DOMContentLoaded') {
            domReadyHandler = handler;
        }
    }
};

const sandbox = {
    window: windowStub,
    document: documentStub,
    Date: Date,
    Math: Math,
    Number: Number,
    String: String,
    Object: Object,
    Array: Array,
    setTimeout: setTimeout,
    clearTimeout: clearTimeout
};

vm.runInNewContext(source, sandbox, {filename: 'info-board-navigation.js'});

const api = windowStub.RssInfoBoardNavigation;
if (!api) {
    throw new Error('RssInfoBoardNavigation API was not exported');
}

const checks = [];
function check(condition, message) {
    checks.push(Boolean(condition));
    console.log((condition ? 'PASS' : 'FAIL') + ': ' + message);
}

check(typeof domReadyHandler === 'function', 'navigation defers DOM initialization until DOMContentLoaded');
check(api.wrappedIndex(0, -1, 5) === 4, 'previous navigation wraps first article to last');
check(api.wrappedIndex(4, 1, 5) === 0, 'next navigation wraps last article to first');
check(api.footerMetaLabel('Example', '8/29 20:10', 2, 10) === 'Example ｜ 8/29 20:10 ｜ 3 / 10',
      'footer metadata combines source, date and position');

let currentText = 'same';
let writeCount = 0;
const textNode = {};
Object.defineProperty(textNode, 'textContent', {
    get: function () { return currentText; },
    set: function (value) { currentText = String(value); writeCount += 1; }
});

check(api.setTextIfChanged(textNode, 'same') === false && writeCount === 0,
      'identical footer text causes no DOM write');
check(api.setTextIfChanged(textNode, 'changed') === true && writeCount === 1 && currentText === 'changed',
      'changed footer text is written exactly once');
check(api.setTextIfChanged(textNode, 'changed') === false && writeCount === 1,
      'repeated footer refresh does not rewrite unchanged text');

const footerTarget = {
    closest: function (selector) {
        return selector === '.info-board-footer' ? {} : null;
    }
};
const outsideTarget = {
    closest: function () { return null; }
};

check(api.dashboardMutationNeedsRefresh([{type: 'childList', target: footerTarget}]) === false,
      'footer-only child mutation is ignored by the global dashboard observer');
check(api.dashboardMutationNeedsRefresh([{type: 'childList', target: outsideTarget}]) === true,
      'non-footer child mutation still refreshes Information Board bindings');
check(api.dashboardMutationNeedsRefresh([{type: 'attributes', target: footerTarget}]) === true,
      'relevant attribute mutations are not suppressed accidentally');
check(api.dashboardMutationNeedsRefresh([]) === true,
      'empty/unknown observer batches remain conservative');

const failed = checks.length - checks.filter(Boolean).length;
console.log('RESULT: PASS ' + (checks.length - failed) + ' / FAIL ' + failed + ' / SKIP 0');
if (failed > 0) {
    process.exit(1);
}

File: tests/test_v1_26_d_info_board_navigation.py
#!/usr/bin/env python3
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
backend = (ROOT / 'app/info_board.php').read_text(encoding='utf-8')
nav = (ROOT / 'public/js/info-board-navigation.js').read_text(encoding='utf-8')
css = (ROOT / 'public/css/info-board.css').read_text(encoding='utf-8')
loader = (ROOT / 'public/js/memo-counter.js').read_text(encoding='utf-8')
runner = (ROOT / 'tests/run-current-features.sh').read_text(encoding='utf-8')

checks: list[bool] = []


def check(ok: bool, message: str) -> None:
    checks.append(bool(ok))
    print(('PASS' if ok else 'FAIL') + ': ' + message)


check('INFO_BOARD_SUMMARY_HARD_LIMIT = 3000' in backend,
      'Information Board keeps a bounded 3000-character RSS summary ceiling')
check("$displayConfig['summary_max'] = INFO_BOARD_SUMMARY_HARD_LIMIT" in backend,
      'runtime Information Board fetch expands legacy summary settings to the safe ceiling')
check("info_board_text_length($content) > info_board_text_length($summary)" in backend,
      'richer RSS content is preferred when it carries more text than description')
check('api_safe_feed_payload($rawFeed, $effectiveUrl)' in backend,
      'summary expansion remains inside the existing sanitized RSS payload boundary')
check('FeedFetchService::fromRuntimeConfiguration()' in backend,
      'summary expansion does not add an article-page fetch path')
check("info-board-nav-prev" in nav and "info-board-nav-next" in nav,
      'previous and next article controls are rendered')
check("前の記事へ戻る" in nav and "次の記事へ進む" in nav,
      'article navigation controls have accessible Japanese labels')
check('wrappedIndex(current, delta, length)' in nav and '% length' in nav,
      'manual navigation wraps between first and last articles')
check('state.index = target' in nav and 'state.needsRestart = true' in nav,
      'manual navigation changes the active ticker article and restarts its summary')
check('INTERACTION_RESUME_DELAY = 5000' in nav and 'state.interactionUntil = Date.now()' in nav,
      'manual article navigation pauses automatic motion before resume')
check("querySelector('.info-board-item-meta')" in nav,
      'footer reuses already sanitized source/date metadata from the rendered article')
check("parts.push(String(Number(index || 0) + 1) + ' / '" in nav,
      'footer shows the current article position')
check("parts.join(' ｜ ')" in nav,
      'site, date, and position are separated readably in the footer')
check("aria-live', 'off'" in nav,
      'automatic article changes do not repeatedly announce footer metadata')
check("info-board-footer" in css and "margin-top: auto" in css,
      'footer uses the lower card space instead of leaving a large empty gap')
check("info-board-footer-meta" in css and "text-overflow: ellipsis" in css,
      'long site metadata stays contained within the card')
check("info-board-nav-button" in css and "min-width: 44px" in css and "min-height: 44px" in css,
      'mobile article navigation targets remain at least 44px')
check("info-board-navigation.js' + assetQuery" in loader,
      'dashboard bootstrap loads the new Information Board navigation asset')
check("data-info-board-v126d-navigation-script" in loader,
      'navigation asset is injected only once')
check("summaryWrap.hidden !== true" in nav and "summaryWrap.hidden = true" in nav,
      'obsolete summary-length selector is hidden without repeating the same DOM write')
check("function setTextIfChanged" in nav and "String(node.textContent || '') === value" in nav,
      'footer text writes are idempotent and skipped when content is unchanged')
check("setTextIfChanged(meta, label)" in nav and "setTextIfChanged(meta, '')" in nav,
      'footer metadata never rewrites identical text on each observer pass')
check("function dashboardMutationNeedsRefresh" in nav and "mutationIsInsideFooter" in nav,
      'dashboard observer can identify self-generated footer mutations')
check("if (!dashboardMutationNeedsRefresh(records))" in nav,
      'global observer ignores footer-only child mutations before re-preparing cards')
check('node --check "$ROOT/public/js/info-board-navigation.js"' in runner,
      'active current-feature runner syntax-checks navigation JavaScript')
check('test_v1_26_d_info_board_navigation.py' in runner,
      'active current-feature runner executes the navigation/footer static contract')
check('test_v1_26_d_info_board_navigation.js' in runner,
      'active current-feature runner executes the observer-loop runtime helper test')
check('fetch(' not in nav and 'XMLHttpRequest' not in nav and '$.ajax' not in nav,
      'navigation/footer adds no network or article scraping path')
check('localStorage' not in nav and 'sessionStorage' not in nav,
      'manual navigation state remains temporary and is not persisted')

failed = len(checks) - sum(checks)
print(f'RESULT: PASS {sum(checks)} / FAIL {failed} / SKIP 0')
raise SystemExit(1 if failed else 0)
